src/entities/addLiquidity/addLiquidityCowAmm/index.ts

```diff
@@ -0,0 +1,135 @@
+    ): Promise<AddLiquidityBaseQueryOutput> {
+        const sortedTokens = getSortedTokens(poolState.tokens, input.chainId);
+
+        const client = createPublicClient({
+            transport: http(input.rpcUrl),
+            chain: CHAINS[input.chainId],
+        });
+
+        const balances: bigint[] = [];
+        for (const token of sortedTokens) {
+            balances.push(
+                await getPoolTokenBalanceCowAmm(
+                    poolState.id,
+                    token.address,
+                    client,
+                ),
+            );
+        }
+
+        const totalShares = await getTotalSupplyCowAmm(
+            poolState.address,
+            client,
+        );
+
+        const poolStateWithBalances: PoolStateWithBalances = {
+            ...poolState,
+            tokens: sortedTokens.map((token, i) => ({
+                address: token.address,
+                decimals: token.decimals,
+                index: i,
+                balance: formatUnits(
+                    balances[i],
+                    token.decimals,
+                ) as HumanAmount,
+            })),
+            totalShares: formatEther(totalShares) as HumanAmount,
+        };
+
+        const { tokenAmounts, bptAmount } = calculateProportionalAmounts(
+            poolStateWithBalances,
+            input.bptOut,
+        );
+        const bptOut = TokenAmount.fromRawAmount(
+            new Token(input.chainId, bptAmount.address, bptAmount.decimals),
+            bptAmount.rawAmount,
+        );
+        const amountsIn = tokenAmounts.map((amountIn) =>
+            TokenAmount.fromRawAmount(
+                new Token(input.chainId, amountIn.address, amountIn.decimals),
+                amountIn.rawAmount,
+            ),
+        );
+        const tokenInIndex = undefined;
+
+        const output: AddLiquidityBaseQueryOutput = {
+            poolType: poolStateWithBalances.type,
+            poolId: poolStateWithBalances.id,
+            addLiquidityKind: input.kind,
+            bptOut,
+            amountsIn,
+            tokenInIndex,
+            chainId: input.chainId,
+            vaultVersion: 0,
+        };
+
+        return output;
+    }
+
+    buildCall(
+        input: AddLiquidityBaseBuildCallInput,
+    ): AddLiquidityBuildCallOutput {
+        if (input.addLiquidityKind !== AddLiquidityKind.Proportional) {
+            throw new Error(
+                `Error: Add Liquidity ${input.addLiquidityKind} is not supported. Cow AMM pools support Add Liquidity Proportional only.`,
+            );
+        }
+
+        const amounts = getAmountsCall(input);
+        const callData = encodeFunctionData({
+            abi: cowAmmPoolAbi,
+            functionName: 'joinPool',
+            args: [amounts.minimumBpt, amounts.maxAmountsInWithoutBpt],
+        });
+
+        return {
+            callData,
+            to: input.poolId,
+            value: getValue(input.amountsIn, !!input.wethIsEth),
```

Comment:
I think the conclusion was native asset is not supported? In that case we should probably check and throw early if it has been incorrectly configured by consumer?

Reply:
done ✅